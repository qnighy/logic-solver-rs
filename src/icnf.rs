use std::collections::HashMap;

use crate::prop::{Id, IdGen, Prop};

/// iCNF (implicational CNF), a modified version of CNF for intuitionistic logic.
///
/// Traditional CNF can be seen as the following sequent:
///
/// ```text
/// (a \/ ~b), (b \/ ~c), (c \/ ~d \/ e) |- false
/// ```
///
/// In this interpretation, CNF-SAT's goal is to **refute** the sequent.
///
/// We extend the "CNF" above in the following ways:
///
/// - Unlike classical CNF, there is a succedent (conclusion part of a sequent).
///   iCNF's succedent is always a variable.
/// - Negative literals are treated differently. Instead of `a \/ b \/ ~c \/ ~d`, we use `c /\ d -> a \/ b` as a clause.
///   Note that they're classically equivalent, but intuitionistically different.
/// - There is a special type of clause in the form of `(a -> b) -> c`.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Icnf {
    /// Hypotheses
    pub ant: Vec<Clause>,
    /// Goal
    pub suc: Id,
}

impl Icnf {
    pub fn from_prop(idgen: &mut IdGen, prop: &Prop) -> (Icnf, Decomposition) {
        let mut ant = vec![];
        let mut decomp = Decomposition {
            map: HashMap::new(),
        };
        let suc = Self::from_prop_with(idgen, &mut decomp, &mut ant, prop, true, false);
        (Self { ant, suc }, decomp)
    }
    fn from_prop_with(
        idgen: &mut IdGen,
        decomp: &mut Decomposition,
        ant: &mut Vec<Clause>,
        prop: &Prop,
        positive: bool,
        negative: bool,
    ) -> Id {
        match prop {
            Prop::Atom(id) => {
                decomp.map.insert(*id, ShallowProp::Atom);
                *id
            }
            Prop::Impl(lhs, rhs) => {
                let lhs = Self::from_prop_with(idgen, decomp, ant, lhs, negative, positive);
                let rhs = Self::from_prop_with(idgen, decomp, ant, rhs, positive, negative);
                let id = idgen.fresh();
                if positive {
                    ant.push(Clause::Impl(lhs, rhs, id));
                }
                if negative {
                    ant.push(Clause::Conj(vec![id, lhs], rhs));
                }
                decomp.map.insert(id, ShallowProp::Impl(lhs, rhs));
                id
            }
            Prop::Conj(children) => {
                let children = children
                    .iter()
                    .map(|child| {
                        Self::from_prop_with(idgen, decomp, ant, child, positive, negative)
                    })
                    .collect::<Vec<_>>();
                let id = idgen.fresh();
                if positive {
                    ant.push(Clause::Conj(children.clone(), id));
                }
                if negative {
                    for &child in &children {
                        ant.push(Clause::Conj(vec![id], child));
                    }
                }
                decomp.map.insert(id, ShallowProp::Conj(children));
                id
            }
            Prop::Disj(children) => {
                let children = children
                    .iter()
                    .map(|child| {
                        Self::from_prop_with(idgen, decomp, ant, child, positive, negative)
                    })
                    .collect::<Vec<_>>();
                let id = idgen.fresh();
                if positive {
                    for &child in &children {
                        ant.push(Clause::Conj(vec![child], id));
                    }
                }
                if negative {
                    ant.push(Clause::Disj(vec![id], children.clone()));
                }
                decomp.map.insert(id, ShallowProp::Disj(children));
                id
            }
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Clause {
    /// (p1 & p2 & ... & pn) -> q
    Conj(Vec<Id>, Id),
    /// (p1 & p2 & ... & pn) -> (q1 | q2 | ... | qn)
    Disj(Vec<Id>, Vec<Id>),
    /// (p -> q) -> r
    Impl(Id, Id, Id),
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Decomposition {
    map: HashMap<Id, ShallowProp>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ShallowProp {
    Atom,
    Impl(Id, Id),
    Conj(Vec<Id>),
    Disj(Vec<Id>),
}

#[cfg(test)]
mod tests {
    use super::*;
    use maplit::hashmap;

    #[test]
    fn test_from_prop1() {
        let mut idgen = IdGen::new();
        let id1 = idgen.fresh();
        let prop = Prop::Atom(id1);
        let result = Icnf::from_prop(&mut idgen, &prop);
        assert_eq!(
            result,
            (
                Icnf {
                    ant: vec![],
                    suc: Id(0)
                },
                Decomposition {
                    map: hashmap! {Id(0) => ShallowProp::Atom},
                }
            )
        );
    }

    #[test]
    fn test_from_prop2() {
        let mut idgen = IdGen::new();
        let id1 = idgen.fresh();
        let prop = Prop::Impl(Box::new(Prop::Atom(id1)), Box::new(Prop::Atom(id1)));
        let result = Icnf::from_prop(&mut idgen, &prop);
        assert_eq!(
            result,
            (
                Icnf {
                    ant: vec![Clause::Impl(Id(0), Id(0), Id(1)),],
                    suc: Id(1)
                },
                Decomposition {
                    map: hashmap! {
                        Id(0) => ShallowProp::Atom,
                        Id(1) => ShallowProp::Impl(Id(0), Id(0)),
                    },
                }
            )
        );
    }

    #[test]
    fn test_from_prop3() {
        let mut idgen = IdGen::new();
        let id1 = idgen.fresh();
        let id2 = idgen.fresh();
        let prop = Prop::Impl(
            Box::new(Prop::Disj(vec![Prop::Atom(id1), Prop::Atom(id2)])),
            Box::new(Prop::Disj(vec![Prop::Atom(id2), Prop::Atom(id1)])),
        );
        let result = Icnf::from_prop(&mut idgen, &prop);
        assert_eq!(
            result,
            (
                Icnf {
                    ant: vec![
                        Clause::Disj(vec![Id(2)], vec![Id(0), Id(1)]),
                        Clause::Conj(vec![Id(1)], Id(3)),
                        Clause::Conj(vec![Id(0)], Id(3)),
                        Clause::Impl(Id(2), Id(3), Id(4)),
                    ],
                    suc: Id(4),
                },
                Decomposition {
                    map: hashmap! {
                        Id(3) => ShallowProp::Disj(vec![Id(1), Id(0)]),
                        Id(1) => ShallowProp::Atom,
                        Id(0) => ShallowProp::Atom,
                        Id(2) => ShallowProp::Disj(vec![Id(0), Id(1)]),
                        Id(4) => ShallowProp::Impl(Id(2), Id(3)),
                    },
                },
            )
        );
    }

    #[test]
    fn test_from_prop4() {
        let mut idgen = IdGen::new();
        let id1 = idgen.fresh();
        let id2 = idgen.fresh();
        let prop = Prop::Impl(
            Box::new(Prop::Conj(vec![Prop::Atom(id1), Prop::Atom(id2)])),
            Box::new(Prop::Conj(vec![Prop::Atom(id2), Prop::Atom(id1)])),
        );
        let result = Icnf::from_prop(&mut idgen, &prop);
        assert_eq!(
            result,
            (
                Icnf {
                    ant: vec![
                        Clause::Conj(vec![Id(2)], Id(0)),
                        Clause::Conj(vec![Id(2)], Id(1)),
                        Clause::Conj(vec![Id(1), Id(0)], Id(3)),
                        Clause::Impl(Id(2), Id(3), Id(4)),
                    ],
                    suc: Id(4),
                },
                Decomposition {
                    map: hashmap![
                        Id(3) => ShallowProp::Conj(vec![Id(1), Id(0)]),
                        Id(1) => ShallowProp::Atom,
                        Id(0) => ShallowProp::Atom,
                        Id(2) => ShallowProp::Conj(vec![Id(0), Id(1)]),
                        Id(4) => ShallowProp::Impl(Id(2), Id(3)),
                    ],
                },
            )
        );
    }
}
