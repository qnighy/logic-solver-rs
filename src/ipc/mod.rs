pub mod icnf;
pub mod icnf_decomp;
pub mod icnf_refutation;
pub mod icnf_solver;
pub mod refutation;
pub mod solver;

pub use refutation::{try_refute, try_refute_res};
pub use solver::{solve, solve_res};
