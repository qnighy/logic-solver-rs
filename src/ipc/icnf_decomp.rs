use std::collections::{HashMap, HashSet};

use crate::ipc::icnf::{ClId, Clause, ClauseSet, Icnf, Var, VarGen};
use crate::prop::{Id, Prop};

#[derive(Debug, Clone)]
pub struct Decomposition {
    sources: HashMap<ClId, ClauseSource>,
    prop_map: PropMap,
}

impl Decomposition {
    pub fn sources(&self) -> &HashMap<ClId, ClauseSource> {
        &self.sources
    }
    pub fn prop_map(&self) -> &PropMap {
        &self.prop_map
    }

    pub fn decompose(vargen: &mut VarGen, prop: &Prop) -> (Icnf, Decomposition) {
        let mut prop_map = PropMap::default();
        let prop_var = prop_map.insert_prop(vargen, prop);

        let mut pol_map = PolarityMap::default();
        pol_map.mark_polarity(&prop_map, prop_var, true);

        let mut sources = HashMap::<ClId, ClauseSource>::new();

        let mut ant = ClauseSet::default();
        for (v, sp) in prop_map.iter() {
            let (pos, neg) = pol_map.polarity(v);
            match sp {
                ShallowProp::Atom(..) => {}
                ShallowProp::Impl(lhs, rhs) => {
                    if pos {
                        let cl = ant.push(Clause::Impl(*lhs, *rhs, v));
                        sources.insert(cl, ClauseSource::ImplIntro(v));
                    }
                    if neg {
                        let cl = ant.push(Clause::Conj(vec![v, *lhs], *rhs));
                        sources.insert(cl, ClauseSource::ImplElim(v));
                    }
                }
                ShallowProp::Conj(children) => {
                    if pos {
                        let cl = ant.push(Clause::Conj(children.clone(), v));
                        sources.insert(cl, ClauseSource::ConjIntro(v));
                    }
                    if neg {
                        for (i, &child) in children.iter().enumerate() {
                            let cl = ant.push(Clause::Conj(vec![v], child));
                            sources.insert(cl, ClauseSource::ConjElim(v, i));
                        }
                    }
                }
                ShallowProp::Disj(children) => {
                    if pos {
                        for (i, &child) in children.iter().enumerate() {
                            let cl = ant.push(Clause::Conj(vec![child], v));
                            sources.insert(cl, ClauseSource::DisjIntro(v, i));
                        }
                    }
                    if neg {
                        let cl = ant.push(Clause::Disj(vec![v], children.clone()));
                        sources.insert(cl, ClauseSource::DisjElim(v));
                    }
                }
            }
        }

        (
            Icnf { ant, suc: prop_var },
            Decomposition { sources, prop_map },
        )
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ClauseSource {
    ImplIntro(Var),
    ImplElim(Var),
    ConjIntro(Var),
    ConjElim(Var, usize),
    DisjIntro(Var, usize),
    DisjElim(Var),
}

#[derive(Debug, Clone, PartialEq, Eq, Default)]
pub struct PropMap {
    rev_vars: HashMap<ShallowProp, Var>,
    vars: Vec<Option<ShallowProp>>,
}

impl PropMap {
    pub fn iter<'a>(&'a self) -> impl Iterator<Item = (Var, &'a ShallowProp)> {
        self.vars
            .iter()
            .enumerate()
            .filter_map(|(i, sp)| Some((Var(i), sp.as_ref()?)))
    }
    pub fn get(&self, v: Var) -> Option<&ShallowProp> {
        self.vars.get(v.0).and_then(|sp| sp.as_ref())
    }
    pub fn insert_prop(&mut self, vargen: &mut VarGen, prop: &Prop) -> Var {
        let sp = match prop {
            Prop::Atom(id) => ShallowProp::Atom(*id),
            Prop::Impl(lhs, rhs) => {
                let lhs = self.insert_prop(vargen, lhs);
                let rhs = self.insert_prop(vargen, rhs);
                ShallowProp::Impl(lhs, rhs)
            }
            Prop::Conj(children) => {
                let children = children
                    .iter()
                    .map(|child| self.insert_prop(vargen, child))
                    .collect::<Vec<_>>();
                ShallowProp::Conj(children)
            }
            Prop::Disj(children) => {
                let children = children
                    .iter()
                    .map(|child| self.insert_prop(vargen, child))
                    .collect::<Vec<_>>();
                ShallowProp::Disj(children)
            }
        };
        self.shallow_insert_prop(vargen, &sp)
    }
    pub fn shallow_insert_prop(&mut self, vargen: &mut VarGen, sp: &ShallowProp) -> Var {
        if let Some(&v) = self.rev_vars.get(sp) {
            v
        } else {
            let v = vargen.fresh();
            self.rev_vars.insert(sp.clone(), v);
            while self.vars.len() <= v.0 {
                self.vars.push(None);
            }
            self.vars[v.0] = Some(sp.clone());
            v
        }
    }

    pub fn to_map(&self) -> HashMap<Var, ShallowProp> {
        self.iter().map(|(v, sp)| (v, sp.clone())).collect()
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Default)]
struct PolarityMap {
    pos: HashSet<Var>,
    neg: HashSet<Var>,
}

impl PolarityMap {
    pub fn polarity(&self, v: Var) -> (bool, bool) {
        (self.pos.contains(&v), self.neg.contains(&v))
    }

    pub fn mark_polarity(&mut self, decomp: &PropMap, v: Var, positive: bool) {
        let polset = if positive {
            &mut self.pos
        } else {
            &mut self.neg
        };
        if polset.contains(&v) {
            return;
        }
        polset.insert(v);
        match decomp.get(v).unwrap() {
            ShallowProp::Atom(..) => {}
            ShallowProp::Impl(lhs, rhs) => {
                self.mark_polarity(decomp, *lhs, !positive);
                self.mark_polarity(decomp, *rhs, positive);
            }
            ShallowProp::Conj(children) => {
                for &child in children {
                    self.mark_polarity(decomp, child, positive);
                }
            }
            ShallowProp::Disj(children) => {
                for &child in children {
                    self.mark_polarity(decomp, child, positive);
                }
            }
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub enum ShallowProp {
    Atom(Id),
    Impl(Var, Var),
    Conj(Vec<Var>),
    Disj(Vec<Var>),
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::prop::IdGen;
    use maplit::hashmap;

    #[test]
    fn test_from_prop1() {
        let mut idgen = IdGen::new();
        let id1 = idgen.fresh();
        let prop = Prop::Atom(id1);
        let (icnf, decomp) = Decomposition::decompose(&mut VarGen::new(), &prop);
        assert_eq!(
            icnf,
            Icnf {
                ant: ClauseSet { vec: vec![] },
                suc: Var(0)
            },
        );
        assert_eq!(
            decomp.prop_map().to_map(),
            hashmap![
                Var(0) => ShallowProp::Atom(Id(0)),
            ]
        );
        assert_eq!(*decomp.sources(), hashmap![]);
    }

    #[test]
    fn test_from_prop2() {
        let mut idgen = IdGen::new();
        let id1 = idgen.fresh();
        let prop = Prop::Impl(Box::new(Prop::Atom(id1)), Box::new(Prop::Atom(id1)));
        let (icnf, decomp) = Decomposition::decompose(&mut VarGen::new(), &prop);
        assert_eq!(
            icnf,
            Icnf {
                ant: ClauseSet {
                    vec: vec![Clause::Impl(Var(0), Var(0), Var(1)),]
                },
                suc: Var(1)
            },
        );
        assert_eq!(
            decomp.prop_map().to_map(),
            hashmap![
                Var(0) => ShallowProp::Atom(Id(0)),
                Var(1) => ShallowProp::Impl(Var(0), Var(0)),
            ],
        );
        assert_eq!(
            *decomp.sources(),
            hashmap![
                ClId(0) => ClauseSource::ImplIntro(Var(1)),
            ]
        );
    }

    #[test]
    fn test_from_prop3() {
        let mut idgen = IdGen::new();
        let id1 = idgen.fresh();
        let id2 = idgen.fresh();
        let prop = Prop::Impl(
            Box::new(Prop::Disj(vec![Prop::Atom(id1), Prop::Atom(id2)])),
            Box::new(Prop::Disj(vec![Prop::Atom(id2), Prop::Atom(id1)])),
        );
        let (icnf, decomp) = Decomposition::decompose(&mut VarGen::new(), &prop);
        assert_eq!(
            icnf,
            Icnf {
                ant: ClauseSet {
                    vec: vec![
                        Clause::Disj(vec![Var(2)], vec![Var(0), Var(1)]),
                        Clause::Conj(vec![Var(1)], Var(3)),
                        Clause::Conj(vec![Var(0)], Var(3)),
                        Clause::Impl(Var(2), Var(3), Var(4)),
                    ]
                },
                suc: Var(4),
            },
        );
        assert_eq!(
            decomp.prop_map().to_map(),
            hashmap! [
                Var(0) => ShallowProp::Atom(Id(0)),
                Var(1) => ShallowProp::Atom(Id(1)),
                Var(2) => ShallowProp::Disj(vec![Var(0), Var(1)]),
                Var(3) => ShallowProp::Disj(vec![Var(1), Var(0)]),
                Var(4) => ShallowProp::Impl(Var(2), Var(3)),
            ],
        );
        assert_eq!(
            *decomp.sources(),
            hashmap![
                ClId(0) => ClauseSource::DisjElim(Var(2)),
                ClId(1) => ClauseSource::DisjIntro(Var(3), 0),
                ClId(2) => ClauseSource::DisjIntro(Var(3), 1),
                ClId(3) => ClauseSource::ImplIntro(Var(4)),
            ]
        );
    }

    #[test]
    fn test_from_prop4() {
        let mut idgen = IdGen::new();
        let id1 = idgen.fresh();
        let id2 = idgen.fresh();
        let prop = Prop::Impl(
            Box::new(Prop::Conj(vec![Prop::Atom(id1), Prop::Atom(id2)])),
            Box::new(Prop::Conj(vec![Prop::Atom(id2), Prop::Atom(id1)])),
        );
        let (icnf, decomp) = Decomposition::decompose(&mut VarGen::new(), &prop);
        assert_eq!(
            icnf,
            Icnf {
                ant: ClauseSet {
                    vec: vec![
                        Clause::Conj(vec![Var(2)], Var(0)),
                        Clause::Conj(vec![Var(2)], Var(1)),
                        Clause::Conj(vec![Var(1), Var(0)], Var(3)),
                        Clause::Impl(Var(2), Var(3), Var(4)),
                    ]
                },
                suc: Var(4),
            },
        );
        assert_eq!(
            decomp.prop_map().to_map(),
            hashmap![
                Var(0) => ShallowProp::Atom(Id(0)),
                Var(1) => ShallowProp::Atom(Id(1)),
                Var(2) => ShallowProp::Conj(vec![Var(0), Var(1)]),
                Var(3) => ShallowProp::Conj(vec![Var(1), Var(0)]),
                Var(4) => ShallowProp::Impl(Var(2), Var(3)),
            ],
        );
        assert_eq!(
            *decomp.sources(),
            hashmap![
                ClId(0) => ClauseSource::ConjElim(Var(2), 0),
                ClId(1) => ClauseSource::ConjElim(Var(2), 1),
                ClId(2) => ClauseSource::ConjIntro(Var(3)),
                ClId(3) => ClauseSource::ImplIntro(Var(4)),
            ]
        );
    }
}
