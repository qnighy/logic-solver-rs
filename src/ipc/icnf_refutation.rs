use std::collections::{HashMap, HashSet};
use varisat::{ExtendFormula, Solver, Var as SatVar};

use crate::ipc::icnf::{Clause, Icnf, Var, VarGen};

const MAX_NUM_WORLDS: usize = 5;

pub fn try_refute_icnf(vargen: &VarGen, icnf: &Icnf) -> Option<Refutation> {
    for num_worlds in 1..=MAX_NUM_WORLDS {
        let mut solver = Solver::new();
        let mapping = register(vargen, icnf, num_worlds, &mut solver);
        let satisfiable = solver.solve().unwrap();
        if satisfiable {
            let model = solver.model().unwrap().into_iter().collect::<HashSet<_>>();
            eprintln!("model = {:?}", model);
            let mut accessibility = HashSet::new();
            for w0 in 0..num_worlds {
                accessibility.insert((w0, w0));
                for w1 in (w0 + 1)..num_worlds {
                    let rel = mapping.wrel[w0 * num_worlds + w1];
                    if model.contains(&rel.positive()) {
                        accessibility.insert((w0, w1));
                    }
                }
            }
            let mut valuation = HashMap::new();
            for i in 0..vargen.max_id() {
                let val = (0..num_worlds)
                    .map(|w| {
                        let sat_var = mapping.varmap[i * num_worlds + w];
                        model.contains(&sat_var.positive())
                    })
                    .collect::<Vec<_>>();
                valuation.insert(Var(i), val);
            }
            return Some(Refutation {
                num_worlds,
                accessibility,
                valuation,
            });
        }
    }
    None
}

#[derive(Debug, Clone, PartialEq, Eq)]
struct RefutationMapping {
    wrel: Vec<SatVar>,
    varmap: Vec<SatVar>,
}

fn register<S: ExtendFormula>(
    vargen: &VarGen,
    icnf: &Icnf,
    num_worlds: usize,
    s: &mut S,
) -> RefutationMapping {
    // accessibility relation
    let mut wrel = vec![SatVar::from_index(0); num_worlds * num_worlds];
    // By symmetry we can assume the accessibility relation is a subrelation of the natural total order.
    // We also assume antisymmetry.
    for i in 0..num_worlds {
        for j in (i + 1)..num_worlds {
            wrel[i * num_worlds + j] = s.new_var();
        }
    }
    // Transitivity
    for i in 0..num_worlds {
        for j in (i + 1)..num_worlds {
            for k in (j + 1)..num_worlds {
                let ij = wrel[i * num_worlds + j];
                let jk = wrel[j * num_worlds + k];
                let ik = wrel[i * num_worlds + k];
                s.add_clause(&[ij.negative(), jk.negative(), ik.positive()]);
            }
        }
    }

    // Valuations
    let mut varmap = vec![SatVar::from_index(0); num_worlds * vargen.max_id()];
    for vi in 0..vargen.max_id() {
        for i in 0..num_worlds {
            varmap[vi * num_worlds + i] = s.new_var();
        }
    }

    // Persistency condition for each variable
    for vi in 0..vargen.max_id() {
        for w0 in 0..num_worlds {
            for w1 in (w0 + 1)..num_worlds {
                let rel = wrel[w0 * num_worlds + w1];
                let vw0 = varmap[vi * num_worlds + w0];
                let vw1 = varmap[vi * num_worlds + w1];
                s.add_clause(&[vw0.negative(), rel.negative(), vw1.positive()]);
            }
        }
    }

    let mut sat_cl = Vec::new();
    let mut impl_var_cache = Vec::new();
    // Encoding clauses
    for (_, cl) in icnf.ant.enumerate() {
        match cl {
            &Clause::Conj(ref lhs, rhs) => {
                for w in 0..num_worlds {
                    sat_cl.clear();
                    for &v in lhs {
                        let sat_v = varmap[v.0 * num_worlds + w];
                        sat_cl.push(sat_v.negative());
                    }
                    let sat_v = varmap[rhs.0 * num_worlds + w];
                    sat_cl.push(sat_v.positive());
                    s.add_clause(&sat_cl);
                }
            }
            &Clause::Disj(ref lhs, ref rhs) => {
                for w in 0..num_worlds {
                    sat_cl.clear();
                    for &v in lhs {
                        let sat_v = varmap[v.0 * num_worlds + w];
                        sat_cl.push(sat_v.negative());
                    }
                    for &v in rhs {
                        let sat_v = varmap[v.0 * num_worlds + w];
                        sat_cl.push(sat_v.positive());
                    }
                    s.add_clause(&sat_cl);
                }
            }
            &Clause::Impl(a, b, c) => {
                impl_var_cache.clear();
                for w0 in 0..num_worlds {
                    sat_cl.clear();
                    for w1 in w0..num_worlds {
                        let sat_var = s.new_var();
                        impl_var_cache.push(sat_var);
                        let sat_aw1 = varmap[a.0 * num_worlds + w1];
                        let sat_bw1 = varmap[b.0 * num_worlds + w1];
                        s.add_clause(&[sat_aw1.positive(), sat_var.positive()]);
                        s.add_clause(&[sat_bw1.negative(), sat_var.positive()]);
                        if w0 < w1 {
                            let rel = wrel[w0 * num_worlds + w1];
                            s.add_clause(&[rel.positive(), sat_var.positive()]);
                        }
                        sat_cl.push(sat_var.negative());
                    }
                    let sat_cw0 = varmap[c.0 * num_worlds + w0];
                    sat_cl.push(sat_cw0.positive());
                    s.add_clause(&sat_cl);
                }
            }
        }
    }

    {
        // Add goal refutation clause
        sat_cl.clear();
        for w in 0..num_worlds {
            let sat_var = varmap[icnf.suc.0 * num_worlds + w];
            sat_cl.push(sat_var.negative());
        }
        s.add_clause(&sat_cl);
    }
    RefutationMapping { wrel, varmap }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Refutation {
    num_worlds: usize,
    accessibility: HashSet<(usize, usize)>,
    valuation: HashMap<Var, Vec<bool>>,
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::ipc::icnf::ClauseSet;
    use insta::{assert_debug_snapshot, assert_snapshot};
    use maplit::hashset;
    use varisat::dimacs::write_dimacs;
    use varisat::CnfFormula;

    #[test]
    fn test_register1() {
        let mut s = CnfFormula::new();
        let mut vargen = VarGen::new();
        let a = vargen.fresh();
        let b = vargen.fresh();
        let ab = vargen.fresh();
        let ba = vargen.fresh();
        let abba = vargen.fresh();
        let icnf = Icnf {
            ant: ClauseSet {
                vec: vec![
                    Clause::Impl(a, b, ab),
                    Clause::Impl(b, a, ba),
                    Clause::Conj(vec![ab], abba),
                    Clause::Conj(vec![ba], abba),
                ],
            },
            suc: abba,
        };
        let mapping = register(&vargen, &icnf, 3, &mut s);
        let dimacs = {
            let mut dimacs = Vec::<u8>::new();
            write_dimacs(&mut dimacs, &s).unwrap();
            String::from_utf8(dimacs).unwrap()
        };
        assert_snapshot!(dimacs);
        assert_debug_snapshot!(mapping);
    }

    #[test]
    fn test_try_refute_icnf1() {
        let mut vargen = VarGen::new();
        let a = vargen.fresh();
        let b = vargen.fresh();
        let ab = vargen.fresh();
        let ba = vargen.fresh();
        let abba = vargen.fresh();
        let icnf = Icnf {
            ant: ClauseSet {
                vec: vec![
                    Clause::Impl(a, b, ab),
                    Clause::Impl(b, a, ba),
                    Clause::Conj(vec![ab], abba),
                    Clause::Conj(vec![ba], abba),
                ],
            },
            suc: abba,
        };
        let refutation = try_refute_icnf(&vargen, &icnf).unwrap();
        assert_eq!(refutation.num_worlds, 3);
        assert_eq!(
            refutation.accessibility,
            hashset![(0, 0), (0, 1), (0, 2), (1, 1), (2, 2)]
        );
        let aval = &refutation.valuation[&a];
        let bval = &refutation.valuation[&b];
        if aval == &[false, false, true] {
            assert_eq!(aval, &[false, false, true]);
            assert_eq!(bval, &[false, true, false]);
        } else {
            assert_eq!(aval, &[false, true, false]);
            assert_eq!(bval, &[false, false, true]);
        }
    }
}
