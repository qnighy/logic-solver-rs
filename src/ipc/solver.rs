use crate::prop::Prop;

use crate::ipc::icnf::VarGen;
use crate::ipc::icnf_decomp::Decomposition;
use crate::ipc::icnf_solver::solve_icnf;
use crate::nj::Proof;

pub fn solve(prop: &Prop) -> Option<Proof> {
    let mut vargen = VarGen::new();
    let (icnf, decomp) = Decomposition::decompose(&mut vargen, prop);
    let icnf_proof = solve_icnf(&vargen, &icnf)?;
    let mut proof = decomp.convert_nj(&icnf_proof, icnf.suc);
    proof.reduce_all();
    Some(proof)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::debruijn::Idx;
    use crate::nj::ProofShorthands;
    use crate::prop::{IdGen, PropShorthands};

    #[test]
    fn test_solve1() {
        use PropShorthands::*;

        let mut idgen = IdGen::new();
        let id1 = idgen.fresh();
        let prop = Atom(id1);
        assert!(solve(&prop).is_none());
    }

    #[test]
    fn test_solve2() {
        use ProofShorthands::*;
        use PropShorthands::*;

        let mut idgen = IdGen::new();
        let id1 = idgen.fresh();
        let a = || Atom(id1);
        let prop = ImplS(a(), a());
        let pf = solve(&prop).unwrap();
        assert_eq!(pf, AbsS(a(), Var(Idx(0))));
    }

    #[test]
    fn test_solve3() {
        use ProofShorthands::*;
        use PropShorthands::*;

        let mut idgen = IdGen::new();
        let id1 = idgen.fresh();
        let id2 = idgen.fresh();
        let a = || Atom(id1);
        let b = || Atom(id2);
        let prop = ImplS(a(), ImplS(b(), a()));
        let pf = solve(&prop).unwrap();
        assert_eq!(pf, AbsS(a(), AbsS(b(), Var(Idx(1)))));
    }
}
