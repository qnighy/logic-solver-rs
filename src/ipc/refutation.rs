use super::icnf::VarGen;
use super::icnf_decomp::Decomposition;
use super::icnf_refutation::try_refute_icnf;
use crate::kripke::KripkeRefutation;
use crate::prop::Prop;

pub fn try_refute(prop: &Prop) -> Option<KripkeRefutation> {
    let mut vargen = VarGen::new();
    let (icnf, decomp) = Decomposition::decompose(&mut vargen, prop);
    let icnf_rft = try_refute_icnf(&vargen, &icnf)?;
    let rft = decomp.convert_kripke(&icnf_rft);
    Some(rft)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::prop::{IdGen, PropShorthands};
    use maplit::hashmap;

    #[test]
    fn test_try_refute1() {
        use PropShorthands::*;

        let mut idgen = IdGen::new();
        let id1 = idgen.fresh();
        let prop = Atom(id1);
        let rft = try_refute(&prop).unwrap();
        assert_eq!(
            rft,
            KripkeRefutation {
                num_worlds: 1,
                accessibility: vec![vec![0]],
                subprops: vec![Atom(id1)],
                valuation: hashmap![
                    Atom(id1) => vec![false],
                ],
            }
        );
    }

    #[test]
    fn test_try_refute2() {
        use PropShorthands::*;

        let mut idgen = IdGen::new();
        let id1 = idgen.fresh();
        let a = || Atom(id1);
        let prop = ImplS(a(), a());
        assert!(try_refute(&prop).is_none());
    }

    #[test]
    fn test_try_refute3() {
        use PropShorthands::*;

        let mut idgen = IdGen::new();
        let id1 = idgen.fresh();
        let id2 = idgen.fresh();
        let a = || Atom(id1);
        let b = || Atom(id2);
        let prop = ImplS(a(), ImplS(b(), a()));
        assert!(try_refute(&prop).is_none());
    }

    #[test]
    fn test_try_refute4() {
        use PropShorthands::*;

        let mut idgen = IdGen::new();
        let id1 = idgen.fresh();
        let a = || Atom(id1);
        let prop = Disj(vec![a(), NegS(a())]);
        let rft = try_refute(&prop).unwrap();
        assert_eq!(
            rft,
            KripkeRefutation {
                num_worlds: 2,
                accessibility: vec![vec![0, 1], vec![1]],
                subprops: vec![a(), Disj(vec![]), NegS(a()), Disj(vec![a(), NegS(a())])],
                valuation: hashmap![
                    a() => vec![false, true],
                    Disj(vec![]) => vec![false, false],
                    NegS(a()) => vec![false, false],
                    Disj(vec![a(), NegS(a())]) => vec![false, true],
                ],
            }
        );
    }
}
