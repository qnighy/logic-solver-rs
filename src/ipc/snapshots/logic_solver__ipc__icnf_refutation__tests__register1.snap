---
source: src/ipc/icnf_refutation.rs
expression: dimacs

---
p cnf 30 59
-1 -3 2 0
-4 -1 5 0
-4 -2 6 0
-5 -3 6 0
-7 -1 8 0
-7 -2 9 0
-8 -3 9 0
-10 -1 11 0
-10 -2 12 0
-11 -3 12 0
-13 -1 14 0
-13 -2 15 0
-14 -3 15 0
-16 -1 17 0
-16 -2 18 0
-17 -3 18 0
4 19 0
-7 19 0
5 20 0
-8 20 0
1 20 0
6 21 0
-9 21 0
2 21 0
-19 -20 -21 10 0
5 22 0
-8 22 0
6 23 0
-9 23 0
3 23 0
-22 -23 11 0
6 24 0
-9 24 0
-24 12 0
7 25 0
-4 25 0
8 26 0
-5 26 0
1 26 0
9 27 0
-6 27 0
2 27 0
-25 -26 -27 13 0
8 28 0
-5 28 0
9 29 0
-6 29 0
3 29 0
-28 -29 14 0
9 30 0
-6 30 0
-30 15 0
-10 16 0
-11 17 0
-12 18 0
-13 16 0
-14 17 0
-15 18 0
-16 -17 -18 0

